File: extensions/vscode/src/extension/services/gitMap.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 alibaba/open-code-review Contributors

import { FileChange } from '@shared/types';
import path from 'path';

export function mapStatusCode(code: string): FileChange['status'] {
  switch (code) {
    case 'A': return 'added';
    case '?': return 'added';
    case 'D': return 'deleted';
    case 'R': return 'renamed';
    case 'M': return 'modified';
    default: return 'modified';
  }
}

/**
 * Parse `git status --porcelain` output.
 * Each line is XY<space>path: X is the index status, Y is the working tree status, and '??' means untracked.
 * Rename lines have the form `R  old -> new`; use the new path.
 */
export function parsePorcelain(output: string): FileChange[] {
  const files: FileChange[] = [];
  const seen = new Set<string>();
  for (const rawLine of output.split('\n')) {
    if (!rawLine.trim()) continue;
    const x = rawLine[0];
    const y = rawLine[1];
    let path = rawLine.slice(3);
    let code: string;
    if (x === '?' && y === '?') {
      code = '?';
    } else if (x === 'R' || y === 'R') {
      code = 'R';
      const arrow = path.indexOf(' -> ');
      if (arrow >= 0) path = path.slice(arrow + 4);
    } else {
      // Prefer the index status, otherwise use the working tree status.
      const c = x !== ' ' && x !== '?' ? x : y;
      code = c;
    }
    path = unquoteGitPath(path);
    if (seen.has(path)) continue;
    seen.add(path);
    files.push({ path, status: mapStatusCode(code) });
  }
  return files;
}

/** Parse untracked paths from `git ls-files --others --exclude-standard` output. */
export function parseUntrackedList(output: string): string[] {
  return output.split('\n').map((line) => unquoteGitPath(line.trim())).filter(Boolean);
}

/** Merge tracked changes with untracked files, deduplicating by path. */
export function mergeWorkspaceFiles(tracked: FileChange[], untrackedPaths: string[]): FileChange[] {
  const files: FileChange[] = [];
  const seen = new Set<string>();
  for (const file of tracked) {
    if (seen.has(file.path)) continue;
    seen.add(file.path);
    files.push(file);
  }
  for (const path of untrackedPaths) {
    if (seen.has(path)) continue;
    seen.add(path);
    files.push({ path, status: 'added' });
  }
  return files;
}

/**
 * Build the workspace file list from git diff and ls-files output.
 * Match OCR CLI workspace mode: use diff HEAD, fall back to staged changes if empty, then merge untracked files.
 */
export function buildWorkspaceFiles(diffHeadOut: string, diffCachedOut: string, untrackedOut: string): FileChange[] {
  let tracked = parseNameStatus(diffHeadOut);
  if (tracked.length === 0) {
    tracked = parseNameStatus(diffCachedOut);
  }
  return mergeWorkspaceFiles(tracked, parseUntrackedList(untrackedOut));
}

/**
 * Choose the candidate repository root that matches the workspace.
 * The VS Code Git extension scans nested repositories asynchronously, so choosing [0] may select a child repository.
 * Priority: exact workspace root match > deepest workspace ancestor > first candidate.
 */
export function pickRepoRoot(roots: string[], workspacePath?: string): string | null {
  if (roots.length === 0) return null;
  if (!workspacePath) return roots[0];

  const candidates = roots.map((root) => {
    const isWindowsPath = [root, workspacePath].some((p) => /^[a-z]:[\\/]/i.test(p) || p.includes('\\'));
    const pathApi = isWindowsPath ? path.win32 : path.posix;
    return { root, relative: pathApi.relative(root, workspacePath), pathApi };
  });

  const exact = candidates.find(({ relative }) => relative === '');
  if (exact) return exact.root;

  const ancestors = candidates.filter(({ relative, pathApi }) =>
    relative !== '..' && !relative.startsWith(`..${pathApi.sep}`) && !pathApi.isAbsolute(relative));
  if (ancestors.length > 0) {
    return ancestors.reduce((deepest, candidate) =>
      candidate.relative.length < deepest.relative.length ? candidate : deepest).root;
  }

  return roots[0];
}

/** Build candidate branch references for rev-parse validation. */
export function branchRefCandidates(ref: string): string[] {
  const candidates = [ref];
  if (!ref.includes('/')) {
    candidates.push(`origin/${ref}`);
  }
  if (ref === 'master') {
    candidates.push('main', 'origin/main');
  } else if (ref === 'main') {
    candidates.push('master', 'origin/master');
  }
  return [...new Set(candidates)];
}

/**
 * Decode Git quotepath escapes (with core.quotepath=true, non-ASCII paths use octal escapes such as "\344\273...").
 */
export function unquoteGitPath(path: string): string {
  if (!path.startsWith('"') || !path.endsWith('"')) return path;

  const bytes: number[] = [];
  const inner = path.slice(1, -1);
  for (let i = 0; i < inner.length; i++) {
    if (inner[i] !== '\\' || i + 1 >= inner.length) {
      bytes.push(inner.charCodeAt(i));
      continue;
    }
    if (i + 3 < inner.length && /^\d{3}$/.test(inner.slice(i + 1, i + 4))) {
      bytes.push(parseInt(inner.slice(i + 1, i + 4), 8));
      i += 3;
      continue;
    }
    i += 1;
    const esc = inner[i];
    if (esc === 'n') bytes.push(0x0a);
    else if (esc === 't') bytes.push(0x09);
    else if (esc === 'r') bytes.push(0x0d);
    else if (esc === '\\') bytes.push(0x5c);
    else if (esc === '"') bytes.push(0x22);
    else bytes.push(esc.charCodeAt(0));
  }
  return Buffer.from(bytes).toString('utf8');
}

/**
 * Parse `git diff --name-status` or `git show --name-status` output.
 * Lines are tab-separated: status<TAB>path, or R<score><TAB>old<TAB>new for renames (use new).
 */
export function parseNameStatus(output: string): FileChange[] {
  const files: FileChange[] = [];
  const seen = new Set<string>();
  for (const rawLine of output.split('\n')) {
    if (!rawLine.trim()) continue;
    const parts = rawLine.split('\t');
    if (parts.length < 2) continue;
    const codeChar = parts[0][0];
    const path = unquoteGitPath(parts.length >= 3 ? parts[parts.length - 1] : parts[1]);
    if (seen.has(path)) continue;
    seen.add(path);
    files.push({ path, status: mapStatusCode(codeChar) });
  }
  return files;
}

File: extensions/idea/src/main/kotlin/com/alibaba/opencodereview/idea/services/GitMap.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 alibaba/open-code-review Contributors

package com.alibaba.opencodereview.idea.services

import com.alibaba.opencodereview.idea.model.FileChange
import com.alibaba.opencodereview.idea.model.FileStatus

/**
 * All pure functions (git text -> domain models) with no IDE API dependencies, so they can be unit-tested directly.
 */

fun mapStatusCode(code: Char): FileStatus = when (code) {
    'A' -> FileStatus.ADDED
    '?' -> FileStatus.ADDED
    'D' -> FileStatus.DELETED
    'R' -> FileStatus.RENAMED
    'M' -> FileStatus.MODIFIED
    else -> FileStatus.MODIFIED // C (copied) / T (typechange) / U (unmerged) and the like all degrade to MODIFIED, see GitMapTest
}

/**
 * Parses `git status --porcelain` output.
 * Each line is XY<space>path, where X is the staged status, Y the worktree status, and `??` means untracked.
 * A rename line looks like `R  old -> new`; take new.
 */
fun parsePorcelain(output: String): List<FileChange> {
    val files = mutableListOf<FileChange>()
    val seen = mutableSetOf<String>()
    for (rawLine in output.lineSequence()) {
        if (rawLine.isBlank() || rawLine.length <= 3) continue
        val x = rawLine[0]
        val y = rawLine[1]
        var path = rawLine.substring(3)
        val code: Char
        if (x == '?' && y == '?') {
            code = '?'
        } else if (x == 'R' || y == 'R' || x == 'C' || y == 'C') {
            // Both renames (R) and copies (C) use the `old -> new` format; take new.
            code = if (x == 'R' || y == 'R') 'R' else 'C'
            // Quoted form (the old path contains spaces or other special characters, so git quoted it): the
            // separator is `" -> ` (closing quote + arrow).
            // Do not search for `" -> "` (one quote more) -- the new path may be unquoted (no special characters),
            // so the new path cannot be required to carry quotes either.
            val quotedSep = path.indexOf("\" -> ")
            if (quotedSep >= 0) {
                path = path.substring(quotedSep + 5) // skip `" -> ` (5 characters), keeping the new path
            } else {
                // Unquoted form: the file name has no spaces, so ` -> ` cannot appear inside it and a plain search is safe.
                val arrow = path.indexOf(" -> ")
                if (arrow >= 0) path = path.substring(arrow + 4)
            }
        } else {
            // Staged status takes priority; fall back to the worktree status when there is none
            code = if (x != ' ' && x != '?') x else y
        }
        path = unquoteGitPath(path)
        if (!seen.add(path)) continue
        files += FileChange(path, mapStatusCode(code))
    }
    return files
}

/** Parses the untracked path list produced by `git ls-files --others --exclude-standard`. */
fun parseUntrackedList(output: String): List<String> =
    output.lineSequence().map { unquoteGitPath(it.trim()) }.filter(String::isNotEmpty).toList()

/** Merges tracked changes with untracked files, deduplicating by path with tracked entries taking priority. */
fun mergeWorkspaceFiles(tracked: List<FileChange>, untrackedPaths: List<String>): List<FileChange> {
    val files = mutableListOf<FileChange>()
    val seen = mutableSetOf<String>()
    for (file in tracked) {
        if (seen.add(file.path)) files += file
    }
    for (path in untrackedPaths) {
        if (seen.add(path)) files += FileChange(path, FileStatus.ADDED)
    }
    return files
}

/**
 * Builds the workspace file list, matching the OCR CLI's workspace mode:
 * `diff HEAD` first, falling back to `diff --cached` when it is empty (no HEAD before the first commit),
 * then merging in the untracked files.
 */
fun buildWorkspaceFiles(diffHeadOut: String, diffCachedOut: String, untrackedOut: String): List<FileChange> {
    var tracked = parseNameStatus(diffHeadOut)
    if (tracked.isEmpty()) tracked = parseNameStatus(diffCachedOut)
    return mergeWorkspaceFiles(tracked, parseUntrackedList(untrackedOut))
}

/**
 * Parses `git branch -a --format=%(refname)` output. The full refname is used deliberately instead of
 * `%(refname:short)`: git abbreviates `refs/remotes/origin/HEAD` to `origin`, which neither ends in HEAD nor
 * equals HEAD, so post-filtering cannot catch it and the dropdown would show a fake branch that errors with
 * "unknown revision" when selected. The full refname allows precisely excluding the `refs/remotes/<remote>/HEAD`
 * symbolic refs.
 */
fun parseBranchList(output: String): List<String> {
    val branches = mutableListOf<String>()
    val seen = mutableSetOf<String>()
    for (rawLine in output.lineSequence()) {
        val ref = rawLine.trim()
        if (ref.isEmpty()) continue
        val name = when {
            ref.startsWith("refs/heads/") -> ref.removePrefix("refs/heads/")
            ref.startsWith("refs/remotes/") -> ref.removePrefix("refs/remotes/")
            // Refs other than heads/remotes (tags, stash, ...) should not appear in branch -a output; skip them defensively.
            ref.startsWith("refs/") -> continue
            else -> ref
        }
        if (name.isEmpty() || name == "HEAD") continue
        if (seen.add(name)) branches += name
    }
    return branches
}

/**
 * Moves `origin/HEAD` and the default remote branch it points at ([defaultRemote]) to the front of the list,
 * right after the first local branch, keeping the remaining entries in their original order (matching the
 * branch order on GitHub's web UI). When [defaultRemote] is null (the repository has no origin/HEAD), the list
 * is returned as-is with no reordering. The positions align with the VS Code extension: local default branch,
 * `origin/HEAD`, default remote branch, then the rest.
 */
fun pinDefaultBranches(branches: List<String>, defaultRemote: String?): List<String> {
    if (defaultRemote.isNullOrBlank()) return branches
    val pin = listOf("origin/HEAD", defaultRemote).filter { it in branches }.distinct()
    if (pin.isEmpty()) return branches
    val rest = branches.filter { it !in pin }
    if (rest.isEmpty()) return pin
    return listOf(rest.first()) + pin + rest.drop(1)
}

/** Builds branch ref candidates for `rev-parse --verify`: add the origin/ prefix and swap main with master. */
fun branchRefCandidates(ref: String): List<String> {
    val candidates = mutableListOf(ref)
    if (!ref.contains('/')) candidates += "origin/$ref"
    when (ref) {
        "master" -> candidates += listOf("main", "origin/main")
        "main" -> candidates += listOf("master", "origin/master")
    }
    return candidates.distinct()
}

/**
 * Decodes git's quotepath escapes (with `core.quotepath=true`, Chinese paths become `"\344\275\240"`).
 * Every git invocation passes `-c core.quotepath=false`, but repository/global config can still introduce
 * the quoted form, so this decoding layer is kept.
 */
fun unquoteGitPath(path: String): String {
    if (path.length < 2 || !path.startsWith('"') || !path.endsWith('"')) return path

    val bytes = mutableListOf<Byte>()
    val inner = path.substring(1, path.length - 1)
    var i = 0
    while (i < inner.length) {
        val ch = inner[i]
        if (ch != '\\' || i + 1 >= inner.length) {
            bytes += ch.code.toByte()
            i++
            continue
        }
        // Octal escape: the digit range is tightened to 0-7 so toInt(8) does not throw on '8'/'9'.
        if (i + 3 < inner.length && inner.substring(i + 1, i + 4).all { it in '0'..'7' }) {
            bytes += inner.substring(i + 1, i + 4).toInt(8).toByte()
            i += 4
            continue
        }
        i++
        when (val esc = inner[i]) {
            'n' -> bytes += 0x0a.toByte()
            't' -> bytes += 0x09.toByte()
            'r' -> bytes += 0x0d.toByte()
            '\\' -> bytes += 0x5c.toByte()
            '"' -> bytes += 0x22.toByte()
            else -> bytes += esc.code.toByte()
        }
        i++
    }
    return String(bytes.toByteArray(), Charsets.UTF_8)
}

/**
 * Parses `git diff --name-status` / `git show --name-status` output.
 * Tab-separated: `status<TAB>path`; a rename is `R<score><TAB>old<TAB>new`, take new.
 */
fun parseNameStatus(output: String): List<FileChange> {
    val files = mutableListOf<FileChange>()
    val seen = mutableSetOf<String>()
    for (rawLine in output.lineSequence()) {
        if (rawLine.isBlank()) continue
        val parts = rawLine.split('\t')
        if (parts.size < 2) continue
        val codeChar = parts[0].firstOrNull() ?: continue
        val path = unquoteGitPath(if (parts.size >= 3) parts.last() else parts[1])
        if (!seen.add(path)) continue
        files += FileChange(path, mapStatusCode(codeChar))
    }
    return files
}
